src/java/org/apache/cassandra/index/sai/disk/v1/SegmentMetadata.java

```diff
@@ -95,7 +95,7 @@ public SegmentMetadata(long segmentRowIdOffset,
                            ByteBuffer maxTerm,
                            ComponentMetadataMap componentMetadatas)
     {
-        assert numRows < Integer.MAX_VALUE;
+        // numRows can exceed Integer.MAX_VALUE because it is the count of unique term and segmentRowId pairs.
```

Comment:
Because of this assertion, I posit that this PR has the same "risk" of int overflow and is therefore an acceptable change. Essentially, I moved this assertion from the `SegmentMetadata` constructor to the `RowMapping` class with the following exception

```java
throw new IllegalArgumentException("RowId must be less than or equal to Integer.MAX_VALUE");
```